### ISSUE TITLE
ActiveModel exclusivity validation doesn't work when value is nil (rails console)

### ISSUE THREAD

**Issue by joemsak** [rails/rails#25461]
### Steps to reproduce

Create a model that validates exclusivity, with a lambda as your `in:` condition:

```ruby
class MyModel < ActiveRecord::Base
  validates :some_field, exclusivity: { in: ->(my_model) { my_model.some_reference } }

  def some_reference
    "reserved word"
  end
end
```

Use rails console to create a new record:

```
bash> rails c
rails-console> my_model = MyModel.new # some_field is nil
rails-console> my_model.valid?
TypeError: no implicit conversion of nil into String
from /path/to/gems/ruby/2.3.0/gems/activemodel-4.2.6/lib/active_model/validations/clusivity.rb:26:in `include?'
```

I'm guessing this is because in the general case the field will be a blank string from web forms, but it did throw me off trying to work in console for a while.

### Expected behavior
The model should validate since `nil` is not in the exclusion list

### Actual behavior
Ruby raises an exception because `nil` cannot be passed to `include?`

### System configuration
**Rails version**: 4.2.6

**Ruby version**: 2.3.1 (my gem path says the gem is installed for 2.3.0 but my shell says I'm using 2.3.1)

**Comment by maclover7** [comment]
Hmm, my attempt at a reproduction script [here](https://gist.github.com/maclover7/ec288a5aa4ea662e28f9e6d0b2aa42d4) passes on both `4-2-stable` and `master`. Can you create a reproduction script that reproduces the error for you?

**Comment by joemsak** [comment]
Wow, it must be something about my setup... I'll look more deeply into it, and let you know asap, thank you

**Comment by joemsak** [comment]
Ah, okay I'm sorry, in my specific example, it's because I am using a lambda for the `in:` option.

Sorry, I assumed it would be a problem either way, since `value` comes from the field being validated, and not the `in:` test.

https://gist.github.com/joemsak/aec697f1bde023bf02039890aac7a799

**Comment by joemsak** [comment]
I updated my initial description of this issue.

**Comment by maclover7** [comment]
Thanks for updating!

Not sure what the best course of action is here. Technically, if users are providing their own lambdas to `in:`, I'd think that it would be the user's responsibility to make sure that `true` or `false` would be returned, and Active Model shouldn't have to rescue from any possible errors, and default to `false`. 😬

**Comment by joemsak** [comment]
Excuse me if I'm wrong in my assumption, but isn't the problem that `value` (which comes from the field being validated) is nil?

This has nothing to do with the lambda/proc returning true or false, from what I can tell.

It doesn't pose a problem when using webforms because the value will be an empty string, but when working in the console and trying to test things, it raises an exception.

**Comment by joemsak** [comment]
In other words, from working at line 26 in a pry, I can't see how it's my the lambda/proc that is causing the nil.

**Comment by maclover7** [comment]
Yes, the issue is that nil is being passed in, but it isn't so much the `nil` itself, as much as what's being acted on the `nil`. In the `in` statement (no pun intended), it checks to see if `nil` is included in the string. Here's that same code, but in just regular `irb`:

```
Jonathans-MBP:rails jon$ irb
irb(main):001:0> 'hi'.include?(nil)
TypeError: no implicit conversion of nil into String
	from (irb):1:in `include?'
	from (irb):1
	from /Users/jon/.rbenv/versions/2.3.0/bin/irb:11:in `<main>'
irb(main):002:0> 
```

Again, the issue isn't that there is a `nil` value, but that the `nil` is being passed into `String#include?` (which is what's getting triggered by the validations here), which seems to not accept `nil` values 😬

**Comment by joemsak** [comment]
Yes, exactly, that is why I opened the issue. It’s my opinion that the
method in this class should handle the nil gracefully


On June 21, 2016 at 1:11:27 PM, Jon Moss (notifications@github.com) wrote:

Yes, the issue is that nil is being passed in, but it isn't so much the nil
itself, as much as what's being acted on the nil. In the in statement (no
pun intended), it checks to see if nil is included in the string. Here's
that same code, but in just regular irb:

Jonathans-MBP:rails jon$ irb
irb(main):001:0> 'hi'.include?(nil)
TypeError: no implicit conversion of nil into String
    from (irb):1:in `include?'
    from (irb):1
    from /Users/jon/.rbenv/versions/2.3.0/bin/irb:11:in `<main>'
irb(main):002:0>

Again, the issue isn't that there is a nil value, but that the nil is being
passed into String#include? (which is what's getting triggered by the
validations here), which seems to not accept nil values 😬

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rails/rails/issues/25461#issuecomment-227524097>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAGCeAdVlWoXbsbjzamPoeR57DOktMLtks5qOClPgaJpZM4I68Ic>
.

**Comment by rafaelfranca** [comment]
This is not an Rails issue. Your lambda should return something that responds to `#include?` and that objects handles `nil` as argument. This works with a Array return for example.

**Comment by joemsak** [comment]
OH I get it now, it’s because my lambda returns a string instead of an array

A-doy! *smacks forehead*

Thank you, your explanation helps!

On June 21, 2016 at 3:45:39 PM, Rafael França (notifications@github.com)
wrote:

Closed #25461 <https://github.com/rails/rails/issues/25461>.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rails/rails/issues/25461#event-699638869>, or mute the
thread
<https://github.com/notifications/unsubscribe/AAGCeKNlLPYwzTekylvRno4S7q3UCN0eks5qOE1ygaJpZM4I68Ic>
.
